File: Week_02/G20190343020145/242.有效的字母异位词.java
import java.util.Arrays;
import java.util.HashMap;

/*
 * @lc app=leetcode.cn id=242 lang=java
 *
 * [242] 有效的字母异位词
 *
 * https://leetcode-cn.com/problems/valid-anagram/description/
 *
 * algorithms
 * Easy (57.15%)
 * Likes:    134
 * Dislikes: 0
 * Total Accepted:    61.9K
 * Total Submissions: 108K
 * Testcase Example:  '"anagram"\n"nagaram"'
 *
 * 给定两个字符串 s 和 t ，编写一个函数来判断 t 是否是 s 的字母异位词。
 * 
 * 示例 1:
 * 
 * 输入: s = "anagram", t = "nagaram"
 * 输出: true
 * 
 * 
 * 示例 2:
 * 
 * 输入: s = "rat", t = "car"
 * 输出: false
 * 
 * 说明:
 * 你可以假设字符串只包含小写字母。
 * 
 * 进阶:
 * 如果输入字符串包含 unicode 字符怎么办？你能否调整你的解法来应对这种情况？
 * 
 */

// @lc code=start
//暴力法，时间复杂度O(nlogn)，空间复杂度O（1）
// class Solution {
//     public boolean isAnagram(String s, String t) {
//         if(s.length() != t.length())
//             return false;
//         char[] ss = s.toCharArray();
//         char[] tt = t.toCharArray();
//         Arrays.sort(ss);
//         Arrays.sort(tt);
//         return Arrays.equals(ss, tt);
//     }
// }

//哈希表 时间复杂度O（n），空间复杂度O（1）
class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length())
            return false;
        //建立一个数组存放hash
        int[] counter = new int[26];

        for(int i = 0; i<s.length();++ i){
            counter[s.charAt(i) - 'a']++;
            counter[t.charAt(i) - 'a']--;
        }
        for(int count : counter) {
            if (count != 0){
                return false;
            }
        }
        return true;

    }
}


// @lc code=end

File: Week_02/G20190343020229/242.有效的字母异位词.py
#
# @lc app=leetcode.cn id=242 lang=python3
#
# [242] 有效的字母异位词
#
# https://leetcode-cn.com/problems/valid-anagram/description/
#
# algorithms
# Easy (56.95%)
# Likes:    137
# Dislikes: 0
# Total Accepted:    63K
# Total Submissions: 109.7K
# Testcase Example:  '"anagram"\n"nagaram"'
#
# 给定两个字符串 s 和 t ，编写一个函数来判断 t 是否是 s 的字母异位词。
# 
# 示例 1:
# 
# 输入: s = "anagram", t = "nagaram"
# 输出: true
# 
# 
# 示例 2:
# 
# 输入: s = "rat", t = "car"
# 输出: false
# 
# 说明:
# 你可以假设字符串只包含小写字母。
# 
# 进阶:
# 如果输入字符串包含 unicode 字符怎么办？你能否调整你的解法来应对这种情况？
# 
#

# @lc code=start
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        smap = {}
        for i in s:
            if i in smap:
                smap[i] += 1
            else:
                smap[i] = 1
        
        for j in t:
            if j in smap:
                smap[j] -= 1
                if smap[j] < 0:
                    return False
            else:
                return False
        return True
        
# @lc code=end

